### PR TITLE
[Reland][FSDP Optim State] Remove checkpoint prefix

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80936

Reland which adds __all__ to torch/utils/checkpoint to silence public binding errors.

### CODE REVIEW

**Review by awgu — APPROVED** [review]

File: test/allowlist_for_publicAPI.json [hunk 1]
```diff
@@ -1960,6 +1960,14 @@
     "Serialization",
     "wrapper_singleton"
   ],
+  "torch.utils.checkpoint": [
```
Comment (anjali411) [inline test/allowlist_for_publicAPI.json:1963]: I am not sure why this is being triggered in your PR ... but anyway this is not the right thing to do. I'd suggest adding an `__all__` in `torch/utils/checkpoint.py` if needed

**Comment by malfet** [comment]
@rohan-varma please avoid `Revert "Revert` commits and at least add some explanation what have changed between this land attempt and the last one

**Comment by rohan-varma** [comment]
@malfet Thanks for letting me know - I updated the PR title and added some info, though unfortunately I guess that won't reflect in the commit history.

**Pull request closed without merging**